src/main/scala/mesosphere/util/CapConcurrentExecutions.scala

```diff
@@ -0,0 +1,157 @@
+  * Only `maxParallel` concurrent executions are allowed.
+  *
+  * {{{
+  * scala> import mesosphere.util.CapConcurrentExecutions
+  * scala> val capParallelExecution = CapConcurrentExecutions(someActorRefFactory, "serialize")
+  * scala> def myFutureReturningFunc: Future[Int] = Future.successful(1)
+  * scala> val result: Future[Int] = capParallelExecution(myFutureReturningFunc)
+  * }}}
+  */
+object CapConcurrentExecutions {
+  private val log = LoggerFactory.getLogger(getClass.getName)
+
+  def apply[T](
+    metrics: CapConcurrentExecutionsMetrics,
+    actorRefFactory: ActorRefFactory,
+    actorName: String,
+    maxParallel: Int,
+    maxQueued: Int): CapConcurrentExecutions = {
+    new CapConcurrentExecutions(metrics, actorRefFactory, actorName, maxParallel, maxQueued)
+  }
+}
+
+class CapConcurrentExecutionsMetrics(metrics: Metrics, metricsClass: Class[_]) {
+  val queued = metrics.gauge(metrics.name(MetricPrefixes.SERVICE, metricsClass, "queued"), new AtomicIntGauge)
+  val processing = metrics.gauge(metrics.name(MetricPrefixes.SERVICE, metricsClass, "processing"), new AtomicIntGauge)
+  val processingTimer = metrics.timer(metrics.name(MetricPrefixes.SERVICE, metricsClass, "processing-time"))
+
+  def reset(): Unit = {
+    queued.setValue(0)
+    processing.setValue(0)
+  }
+}
+
+class CapConcurrentExecutions private (
+    metrics: CapConcurrentExecutionsMetrics,
+    actorRefFactory: ActorRefFactory,
+    actorName: String,
+    maxParallel: Int,
+    maxQueued: Int) {
+  import CapConcurrentExecutions.log
+
+  private[this] val serializeExecutionActorProps =
+    RestrictParallelExecutionsActor.props(metrics, maxParallel = maxParallel, maxQueued = maxQueued)
+  private[this] val serializeExecutionActorRef = actorRefFactory.actorOf(serializeExecutionActorProps, actorName)
+
+  def apply[T](block: => Future[T]): Future[T] = {
+    PromiseActor.askWithoutTimeout(
+      actorRefFactory, serializeExecutionActorRef, RestrictParallelExecutionsActor.Execute(() => block))
+  }
+
+  @PreDestroy
+  def close(): Unit = {
+    log.debug(s"stopping $serializeExecutionActorRef")
+    serializeExecutionActorRef ! PoisonPill
+  }
+}
+
+/**
+  * Accepts execute instructions containing functions returning `scala.concurrent.Future`s.
+  * It only allows `maxParallel` parallel executions and queues the other operations.
+  * It will not queue more than `maxQueued` execute instructions.
+  */
+private[util] class RestrictParallelExecutionsActor(
+    metrics: CapConcurrentExecutionsMetrics, maxParallel: Int, maxQueued: Int) extends Actor {
+
+  import RestrictParallelExecutionsActor.{ log, Execute, Queued }
+
+  private[this] var active: Int = 0
+  private[this] var queue: Queue[Queued] = Queue.empty
+
+  override def preStart(): Unit = {
+    super.preStart()
+
+    metrics.reset()
+
+    for (execute <- queue) {
+      execute.sender ! Status.Failure(new IllegalStateException(s"$self actor stopped"))
```

Comment:
I guess. But since we really should do this in postStop, it doesn't matter that much ;)